### PR TITLE
feat: implement a request cache for mutation and insertion queries

### LINKED ISSUE #137: Implement a request cache
Cache requests to reduce load on SILO.

Details:
- the cache should be invoked for:
    - /nucleotideMutations
    - /aminoAcidMutations
    - /nucleotideInsertions
    - /aminoAcidInsertions
- other routes should be ignored
- when a new data version is available the cache should be invalidated

### PR DESCRIPTION
**Pull request by fengelniederhammer**
Also make sure to invalidate the cache when SILO has a new data version

resolves #137

## PR Checklist
- ~~[ ] All necessary documentation has been adapted.~~
- [x] The implemented feature is covered by an appropriate test.

### CODE REVIEW

File: lapis2/src/main/kotlin/org/genspectrum/lapis/scheduler/DataVersionCacheInvalidator.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+@Component
+class DataVersionCacheInvalidator(
+    private val cachedSiloClient: CachedSiloClient,
+    private val cacheClearer: CacheClearer,
+) {
+    final var currentlyCachedDataVersion = "uninitialized"
+        private set
+
+    @Scheduled(fixedRate = 30, timeUnit = TimeUnit.SECONDS)
```
Comment (fengelniederhammer) [inline lapis2/src/main/kotlin/org/genspectrum/lapis/scheduler/DataVersionCacheInvalidator.kt:21]: @chaoran-chen What's a reasonable interval to check for new versions? I think the old LAPIS does it every second, if I interpret this correctly:
https://github.com/GenSpectrum/LAPIS/blob/831399692fdf1bcce6fc9e4021e950c4ee8bc00f/server/src/main/java/ch/ethz/lapis/api/DataVersionService.java#L40

File: lapis2/src/main/resources/application.properties [hunk 2]
```diff
@@ -2,3 +2,12 @@ springdoc.default-produces-media-type=application/json
 springdoc.api-docs.path=/api-docs
 springdoc.swagger-ui.operationsSorter=alpha
 server.forward-headers-strategy=framework
+
+spring.cache.cache-names=siloQueryCache
+spring.cache.caffeine.spec=maximumSize=50000,expireAfterAccess=1h
```
Comment (fengelniederhammer) [inline lapis2/src/main/resources/application.properties:7]: @chaoran-chen Do you have any experience from LAPIS 1 what reasonable values might be here? As a wild guess, I put 50,000 cached entries at max and a lifetime of 1 hours here.

File: lapis2/src/main/kotlin/org/genspectrum/lapis/scheduler/DataVersionCacheInvalidator.kt (hunk 1 above)
Comment (chaoran-chen) [reply lapis2/src/main/kotlin/org/genspectrum/lapis/scheduler/DataVersionCacheInvalidator.kt:21]: Yes, every second would be good.

File: lapis2/src/main/resources/application.properties (hunk 2 above)
Comment (chaoran-chen) [reply lapis2/src/main/resources/application.properties:7]: I don't have a good number for the number of entries but 50k sounds reasonable. Alternatively, is there a way to restrict the memory size? For LAPIS 1, the max size was set to 2 GB.

I don't think that we need an expiry time. As long as it's the same data version, the results will be the same.

File: lapis2/src/main/kotlin/org/genspectrum/lapis/scheduler/DataVersionCacheInvalidator.kt (hunk 1 above)
Comment (fengelniederhammer) [reply lapis2/src/main/kotlin/org/genspectrum/lapis/scheduler/DataVersionCacheInvalidator.kt:21]: Ok, then I can probably also remove the other logic that checks the data version on a normal query.

**Comment by chaoran-chen** [comment]
@fengelniederhammer, it just came to my mind (in case you haven't considered it already): if the query asks for a random order, the result may not be cached.

File: lapis2/src/main/resources/application.properties (hunk 2 above)
Comment (fengelniederhammer) [reply lapis2/src/main/resources/application.properties:7]: Thinking about it: the larget collection page (that I know of) sends around 2200 requests. Normal variant pages send around 20 relevant requests. So 5000 entries should probably do it here.

**Comment by fengelniederhammer** [comment]
Thanks for the reminder! I added a comment to #658

File: lapis2/src/main/resources/application.properties (hunk 2 above)
Comment (chaoran-chen) [reply lapis2/src/main/resources/application.properties]: Why only 5000? Or do you mean 50,000?

Don't forget that a collection is not static. You can filter by location and time range:

image

Comment (fengelniederhammer) [reply lapis2/src/main/resources/application.properties]: Ah true. Ok, then I'll put it to 50k and no expiry time. Anyway, this config value can be overwritten when starting LAPIS.

Comment (fengelniederhammer) [reply lapis2/src/main/resources/application.properties]: I did not find a way to make the eviction based on memory size. https://github.com/ben-manes/caffeine/wiki/Eviction#size-based

Maybe there are other cache implementations that support it. In principle, it should be easy to exchange. None of the implementation code is specific to the implementation.

Comment (chaoran-chen) [reply lapis2/src/main/resources/application.properties]: sounds good, then let's try with 50k and no expiry time

**Review by JonasKellerer — COMMENTED** [review]
Can we somehow test whether the cache is invalidated when a new data version is available?

Also do we need to add some docu?

**Comment by fengelniederhammer** [comment]
There is `SiloClientAndCacheInvalidatorTest` in `SiloClientTest.kt` that checks tests this: populate the cache with some data version, set the SILO mock to return a new data version and check that the next time, the cache isn't hit.

I'll do the docs in a separate PR.

**Pull request merged**